app/demo-geant-integration/HitRootIO.cc

```diff
@@ -0,0 +1,205 @@
+            {
+                G4VHit* hit = hc->GetHit(j);
+                SensitiveHit* sd_hit = dynamic_cast<SensitiveHit*>(hit);
+                hits.push_back(sd_hit);
+            }
+            hcmap->insert(std::make_pair(hcname, hits));
+        }
+    }
+
+    // Write a HitRootEvent into output ROOT file
+    G4AutoLock lock(&HitRootIOMutex);
+    this->WriteObject(hit_event.release());
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Fill and write a HitRootEvent object
+ */
+void HitRootIO::WriteObject(HitRootEvent* hit_event)
+{
+    if (!init_branch_)
+    {
+        event_branch_ = tree_->Branch(
+            "event.",
+            &hit_event,
+            GlobalSetup::Instance()->GetSetupOptions()->root_buffer_size,
+            this->SplitLevel());
+        init_branch_ = true;
+    }
+
+    tree_->Fill();
+    file_->Write("", TObject::kOverwrite);
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Close or Merge output
+ */
+void HitRootIO::Close()
+{
+    if (!G4Threading::IsMultithreadedApplication())
+    {
+        CELER_LOG(info) << "Writing hit ROOT output to " << file_name_ << "\"";
+        file_ = tree_->GetCurrentFile();
+        file_->Close();
+    }
+    else
+    {
+        // Merge output file on the master thread if MT
+        if (G4Threading::IsMasterThread())
+        {
+            this->Merge();
+        }
+    }
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Merging output root files from multiple threads using TTree::MergeTrees
+ *
+ * TODO: use TBufferMerger and follow the example described in the ROOT
+ * tutorials/multicore/mt103_fillNtupleFromMultipleThreads.C which stores
+ * TBuffer data in memory and writes 32MB compressed output concurrently.
+ */
+void HitRootIO::Merge()
+{
+    auto nthreads = G4RunManager::GetRunManager()->GetNumberOfThreads();
+    std::vector<TFile*> files;
+    std::vector<TTree*> trees;
+    std::unique_ptr<TList> list(new TList);
+
+    celeritas::ExceptionConverter call_g4exception{"celer0006"};
+    CELER_LOG(info) << "Merging hit root files from " << nthreads
+                    << " threads into \"" << file_name_ << "\"";
+
+    for (int i = 0; i < nthreads; ++i)
+    {
+        std::string file_name = file_name_ + std::to_string(i);
+        files.push_back(TFile::Open(file_name.c_str()));
+        trees.push_back((TTree*)(files[i]->Get("Events")));
+        list->Add(trees[i]);
+
+        if (i == nthreads - 1)
+        {
+            TFile* file = TFile::Open(file_name_.c_str(), "recreate");
+            CELER_VALIDATE(file->IsOpen(), << "Failed to open " << file_name_);
+
+            TTree* tree = TTree::MergeTrees(list.get());
+            tree->SetName("Events");
+            auto hit_event = std::make_unique<HitRootEvent>().release();
+            tree->SetBranchAddress("event.", &hit_event);
+            tree->Write();
```

Comment:
```suggestion
            file->Write();
```
As we need to write both the `TFile` and `TTree` meta-data.

Reply:
Changed as suggested.